### PR TITLE
Reduce extension footprint when not in repository context

### LINKED ISSUE #1799: GitHub VS extension is loaded for all C# apps
(Writing from VS platform team)

GitHub extension reads in 2.5 MB off of disk when I create a basic C# app (with just an empty Main method) that doesn't use GitHub. This is with a full install of Visual Studio 15.8 Preview 3.0 (all workloads/packages installed). Disk access especially during solution load can be costly. Therefore, it should be avoided if the project doesn't use a feature. For more information please see [this page](https://msdn.microsoft.com/en-us/library/mt750411.aspx).

Below, please find the breakdown by DLL:

Name | Bytes | Page Ct
-- | -- | --
System.Reactive.Linq.dll (C:\Program   Files (x86)\Microsoft Visual Studio   15.0\Common7\IDE\Extensions\GitHub\GitHub) | 552,960 | 135
GitHub.App.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 319,488 | 78
Octokit.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 270,336 | 66
ReactiveUI.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 225,280 | 55
LibGit2Sharp.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 204,800 | 50
GitHub.InlineReviews.dll (C:\Program   Files (x86)\Microsoft Visual Studio   15.0\Common7\IDE\Extensions\GitHub\GitHub) | 159,744 | 39
Akavache.Sqlite3.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 114,688 | 28
System.Reactive.Core.dll (C:\Program   Files (x86)\Microsoft Visual Studio   15.0\Common7\IDE\Extensions\GitHub\GitHub) | 106,496 | 26
rothko.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 94,208 | 23
SQLitePCL.raw.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 81,920 | 20
Akavache.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 81,920 | 20
GitHub.Exports.Reactive.dll (C:\Program   Files (x86)\Microsoft Visual Studio   15.0\Common7\IDE\Extensions\GitHub\GitHub) | 77,824 | 19
GitHub.TeamFoundation.15.dll (C:\Program   Files (x86)\Microsoft Visual Studio   15.0\Common7\IDE\Extensions\GitHub\GitHub) | 69,632 | 17
Splat.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 57,344 | 14
GitHub.Api.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 40,960 | 10
GitHub.VisualStudio.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 36,864 | 9
System.Reactive.PlatformServices.dll   (C:\Program Files (x86)\Microsoft Visual Studio   15.0\Common7\IDE\Extensions\GitHub\GitHub) | 32,768 | 8
sqlite3.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub\x86) | 32,768 | 8
GitHub.Exports.dll (C:\Program Files   (x86)\Microsoft Visual Studio 15.0\Common7\IDE\Extensions\GitHub\GitHub) | 28,672 | 7
System.Reactive.Windows.Threading.dll   (C:\Program Files (x86)\Microsoft Visual Studio   15.0\Common7\IDE\Extensions\GitHub\GitHub) | 24,576 | 6
GitHub.Services.Vssdk.dll (C:\Program   Files (x86)\Microsoft Visual Studio   15.0\Common7\IDE\Extensions\GitHub\GitHub) | 20,480 | 5
System.Reactive.Interfaces.dll   (C:\Program Files (x86)\Microsoft Visual Studio   15.0\Common7\IDE\Extensions\GitHub\GitHub) | 20,480 | 5

Please do not hesitate to let us know if any further information is needed.

Thank you

### PR DESCRIPTION
**Pull request by jcansdale**
## The problem

@PooyaZv (Writing from VS platform team) wrote:

> GitHub extension reads in 2.5 MB off of disk when I create a basic C# app (with just an empty Main method) that doesn't use GitHub. This is with a full install of Visual Studio 15.8 Preview 3.0 (all workloads/packages installed). Disk access especially during solution load can be costly. Therefore, it should be avoided if the project doesn't use a feature.

We appear to be loading 22 assemblies, even when a project isn't using GitHub (see #1799 for a breakdown).

## What this PR does

1. Only create text view margins when in the context of a Git repository
Change `InlineCommentMarginProvider` and `PullRequestFileMarginProvider` to lazy load their MEF dependencies and only return a margin when in the context of a Git repository.

2. Remove dependency on `LogManager` from `GitContextPackage`
`GitContextPackage` is our custom UI context and is loaded whenever we're in the context of a Git repository. This a low level package and doesn't really need logging (which requires a few assemblies).

3. Change `GitContextPackage` and `VSGitExt` to use `UICONTEXT.RepositoryOpen` instead of `GitSccProvider`
It appears there is a UIContext for when a Git repository is open (`UICONTEXT.RepositoryOpen`). Previously we were using `GitSccProvider`, which loads whenever the Git SCC provider is available (which is most of the time).

4. Stop using `SharedDictionary.xaml` for GitHub logo
The GitHub logo used on `View > Other Windows > GitHub` and `StartPage > GitHub`  was causing a cascade of assembly loads when the `StartPage` or `GitHub` command was visible. I've replaced the `SharedDictionary.xaml` reference with a plain XAML placeholder. This causes only the containing assembly to load.

### Questions

Our custom UIContext is now the same as `UICONTEXT.RepositoryOpen`, but with the added restriction that the host process is Visual Studio (not Blend). I don't know if there is a built in UIContext that would allow us to do the same thing? @PooyaZv do you know?

## To do

- [x] Replace placeholder XAML with proper GitHub logo

## How to test

1. Open Visual Studio with `StartPage` visible and the `GitHub` pane hidden
2. Open the modules window (`Ctrl D + M`) and search for `github`
- Expect only `GitHub.VisualStudio.dll` to have loaded (it contains the GitHub icon)
3. Open a project that isn't under Git version control and open a source file
- Expect `GitHub.InlineReviews.dll`, `GitHub.Exports.dll` and `GitHub.Exports.Reactive.dll` to have loaded (these contain our MEF margin providers and MEF interfaces)
- No other assemblies should have loaded under the extension directory

image

Fixes #1799

### CODE REVIEW

**Comment by jcansdale** [comment]
We appear to be loading a bunch of assemblies when simply viewing a button with the GitHub icon. This impacts both the `View > Other Windows > GitHub` and worse the `StartPage > GitHub` icon.

image

image

**Comment by PooyaZv** [comment]
@jcansdale Thank you :) Can you please add @jialongcheng as a reviewer as well? He has more context on UI contexts than I.

**Comment by jcansdale** [comment]
@PooyaZv, I tried adding @jialongcheng as a reviewer, but for some reason wasn't able to. 😕 

@jialongcheng, do you know if there are any UIContexts I could use to differentiate between the Visual Studio process and Blend? Also, are there UIContexts fir different versions of Visual Studio?

**Comment by jialongcheng** [comment]
@jcansdale we do not have a public UI context for checking if we are in Blend. What is your use case?
We do not have UI context for which version of VS is used.

**Comment by jcansdale** [comment]
@jialongcheng,

> we do not have a public UI context for checking if we are in Blend. What is your use case?

The challenge is actually preventing the extension from running inside Blend. It appears that VS services are disabled inside Blend, but packages and MEF components are still loaded. We need to be very careful not to break when all of our services return null. 😢 

Is there any way to prevent an extension's packages and MEF components from being loaded inside Blend? That would be ideal.

**Comment by jialongcheng** [comment]
@jcansdale Unfortunately we do not have a way at this point to prevent an extension's packages and MEF components from being loaded inside Blend.

**Comment by grokys** [comment]
Trying this out, I get the following on running the experimental instance:

image

This is more than you indicated should be loaded. Is this correct?

This PR definitely does load less assemblies than before when a non-git repo project is opened though! Nice work!

File: src/GitHub.InlineReviews/Margins/InlineCommentMarginProvider.cs [hunk 1]
```diff
@@ -1,11 +1,13 @@
 ﻿using System;
 using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Shell;
```
Comment (StanleyGoldman) [inline src/GitHub.InlineReviews/Margins/InlineCommentMarginProvider.cs]: Test review

File: src/GitHub.InlineReviews/Margins/InlineCommentMarginProvider.cs [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
 using Microsoft.VisualStudio.Utilities;
 using Microsoft.VisualStudio.Text.Editor;
 using Microsoft.VisualStudio.Text.Tagging;
 using Microsoft.VisualStudio.Text.Classification;
-using GitHub.InlineReviews.Services;
 using GitHub.Services;
+using GitHub.VisualStudio;
```
Comment (StanleyGoldman) [inline src/GitHub.InlineReviews/Margins/InlineCommentMarginProvider.cs]: asdf

**Comment by jcansdale** [comment]
> Trying this out, I get the following on running the experimental instance

I've just done some tests and was initially seeing the same as you.

When you launch Visual Studio, make sure the `Start Page`, `Team Explorer` and `GitHub` tool windows are all closed.

If all goes to plan, you should see this:

image

Show the `Start Page`:

image

Open a solution+source file that doesn't use Git:

image
(The InlineReviews project installs its MEF based editor margins)

Open a solution that in on GitHub:

image

Open the `Team Explorer` tool window.  No change.

Open the `GitHub` tool window:

**Comment by jcansdale** [comment]
If you open a repository that is on Git but not GitHub, you'll see this:

image

This isn't great, but maybe something for another PR? It would need to probe for GitHub Enterprise to work properly.

**Pull request merged**